feat: introduce navigation stream

asalzburger opened the pull request:
This PR introduces a `NavigationStream` object and a corresponding helper. 

The `NavigationStream` should build the backbone of the Gen3 navigator, with having a `TargetStream` and a `GeometryStream` to deal with.

For details of the ongoing discussion, please see #3526.


@paulgessinger @andiwand

andiwand commented on Core/include/Acts/Navigation/NavigationStream.hpp:
```diff
@@ -0,0 +1,98 @@
+    const Portal* portal = nullptr;
+    /// The boundary tolerance
+    BoundaryTolerance bTolerance = BoundaryTolerance::None();
+    /// Target flag: true if this Candidate represents an abortTarget
+    bool abortTarget = false;
```
Is this information useful to the navigator? I expected an aborter to know which surfaces trigger an abort and the navigator only to care about approaching them.

asalzburger replied:
Not sure yet, probably not - I needed it for testing behaviour ... will think about it.

asalzburger replied:
Removed.

andiwand commented on Core/include/Acts/Navigation/NavigationStream.hpp:
```diff
@@ -0,0 +1,98 @@
+  /// Non-cost access the current candidate
+  Candidate& currentCandidate() { return candidates.at(currentIndex); }
+
+  /// The number of active candidates
+  std::size_t activeCandidates() const {
+    return (candidates.size() - currentIndex);
+  }
```
alternatively `remainingCandidates`

asalzburger replied:
I like.

andiwand commented on Core/include/Acts/Navigation/NavigationStreamHelper.hpp:
IMO this could be added to the other header. Potentially as static functions on the `NavigationStream` class. To reduce a little the new namespaces and header files we need to include while we are talking about the same concept.

andiwand replied:
Potentially even member functions since we always require a mutable ref

asalzburger replied:
ok, why not.

asalzburger replied:
Done.

andiwand commented on Core/include/Acts/Utilities/Intersection.hpp:
```diff
@@ -145,6 +149,14 @@ class ObjectIntersection {
                                const object_t* object, std::uint8_t index = 0)
       : m_intersection(intersection), m_object(object), m_index(index) {}
 
+  /// Invalid object intersection - only holding the object itself
+  ///
+  /// @param object is the object to be instersected
+  constexpr ObjectIntersection(const object_t* object)
+      : m_intersection(Intersection3D::invalid()),
+        m_object(object),
+        m_index(0) {}
```
There is an `invalid` factory method which would fit more to express the purpose of the object creation

asalzburger replied:
yes, but the invalid does not take a surface, this is the filler method, i.e. you pre-fill your stream with at this moment invalid intersection, but you keep it to hold the surface.

andiwand replied:
Ok I see but the we could have a different overload for `invalid` to keep this consistent. That way we do not introduce another way of creating invalid intersections

asalzburger replied:
I changed it such that the `invalid(...)` could also get an object.

andiwand commented on Core/src/Navigation/NavigationStreamHelper.cpp:
```diff
@@ -0,0 +1,130 @@
+  const Vector3& position = queryPoint.position;
+  const Vector3& direction = queryPoint.direction;
+
+  // De-duplicate first (necessary to deal correctly with multiple
+  // intersections) - sort them by surface pointer
+  std::sort(stream.candidates.begin(), stream.candidates.end(),
+            [](const NavigationStream::Candidate& a,
+               const NavigationStream::Candidate& b) {
+              return (&a.surface()) < (&b.surface());
+            });
+  // Remove duplicates on basis of the surface pointer
+  stream.candidates.erase(
+      std::unique(stream.candidates.begin(), stream.candidates.end(),
+                  [](const NavigationStream::Candidate& a,
+                     const NavigationStream::Candidate& b) {
+                    return (&a.surface()) == (&b.surface());
+                  }),
+      stream.candidates.end());
```
This I don't quite understand. I thought we want to keep multiple intersections so we can potentially cross a surface twice? Thinking of a tube modeled as a cylinder for example we will encounter the surface twice without switching volume.

asalzburger replied:
This is the initialisation call, i.e. you fill 4 cylinders, you get 4 initial candidates.

Then you run the intersections, if you find multiple valid ones per surface, you remember them.
After the intersection run, you join the ones and get all possible intersections.

If you want to re-initialize the same candidates, you need to de-duplicate first and then run the intersection with potential multiplication again.

asalzburger replied:
Moved to `NavigationStream` so that it is callable (if needed).

andiwand commented on Core/src/Navigation/NavigationStreamHelper.cpp:
```diff
@@ -0,0 +1,130 @@
+    const Surface* surface = sIntersection.object();
+    // Intersect the surface
+    auto multiIntersection = surface->intersect(
+        gctx, position, direction, cTolerance, s_onSurfaceTolerance);
```
I think we should have a parameter for the `onSurfaceTolerance` instead of silently using the default one

asalzburger replied:
happy with that.

asalzburger replied:
Fixed.

andiwand commented on Core/src/Navigation/NavigationStreamHelper.cpp:
```diff
@@ -0,0 +1,130 @@
+    bool miCandidate = false;
+    for (auto& rsIntersection : multiIntersection.split()) {
+      // Skip negative solutions
+      if (rsIntersection.pathLength() < 0.) {
```
I usually probe against the `onSurfaceTolerance` instead of `0` because this is the finite space we would still consider to be on surface. Anything beyond that is what we want to encounter

asalzburger replied:
ok, sounds good.

asalzburger replied:
But that should be negative onSurface tolerance, right?

andiwand commented on Core/src/Navigation/NavigationStreamHelper.cpp:
```diff
@@ -0,0 +1,130 @@
+        gctx, position, direction, cTolerance, s_onSurfaceTolerance);
+
+    // Split them into valid intersections
+    bool miCandidate = false;
```
can we spell this out?

asalzburger replied:
yes, will do.

andiwand commented on Core/src/Navigation/NavigationStreamHelper.cpp:
```diff
@@ -0,0 +1,130 @@
+      // Valid solution is either on surface or updates the distance
+      if (rsIntersection.isValid()) {
+        if (!miCandidate) {
+          sIntersection = rsIntersection;
+          miCandidate = true;
```
the names look a bit cryptic to me

asalzburger replied:
agreed.

andiwand commented on Core/src/Navigation/NavigationStreamHelper.cpp:
```diff
@@ -0,0 +1,130 @@
+
+  // Sort the candidates by path length
+  std::sort(stream.candidates.begin(), stream.candidates.end(),
+            [](const NavigationStream::Candidate& a,
+               const NavigationStream::Candidate& b) {
+              return a.intersection.pathLength() < b.intersection.pathLength();
+            });
```
alternatively we could pull in the comparator from the intersection to the `Candidate` class and then supply `pathLengthOrder` here

asalzburger replied:
yes, ok

asalzburger replied:
Done.

andiwand commented on Core/include/Acts/Navigation/NavigationStreamHelper.hpp:
```diff
@@ -0,0 +1,102 @@
+                      const NavigationStream::QueryPoint& queryPoint,
+                      BoundaryTolerance cTolerance);
+
+/// Convenience method to update a stream from a new query point,
+/// this could be called from navigation delegates that do not require
+/// a local state or from the navigator on the target stream
+///
+/// @param stream [in, out] is the navigation stream to be updated
+/// @param gctx is the geometry context
+/// @param queryPoint holds current position, direction, etc.
+///
+/// @return true if the stream is active, false indicate no valid candidates left
+bool updateStream(NavigationStream& stream, const GeometryContext& gctx,
+                  const NavigationStream::QueryPoint& queryPoint);
```
As discussed, I wonder if this has a real application. Reintersecting and sorting is quite expensive and not as accurate as reinitialization so I would just go with reinitialization.

asalzburger replied:
Here, the navigator can use this to update the target stream (in fact, it will update only the next one, and skip over to the next but next only in case the next one is not reached).

Also, the same code can be used in the local LayerStructure stream until exhausted. That's why I put it into a helper.

asalzburger commented:
Hi @andiwand - I've changed this quite a bit, so here's a summary to guide you:

* changed `NavigationStream` into a class with private members
* moved `initialize()` and `update` methods to navigation stream, also the `fill` methods and renamed them
* removed `targetAbort`
* removed `NavigationStreamHelper`

This is pretty much along the lines that you requested.

andiwand reviewed (dismissed):
LGTM 👍 

One question: I wonder if we want to re-converge `Core/Navigation` and `Core/Propagator` since we accumulate stepping, navigation and propagation there. I feel like splitting out `Navigation` requires also to split out `Stepping`.

asalzburger commented:
@andiwand - had a problem with the doc run, so I had to push another commit

asalzburger commented:
> LGTM 👍
> 
> One question: I wonder if we want to re-converge `Core/Navigation` and `Core/Propagator` since we accumulate stepping, navigation and propagation there. I feel like splitting out `Navigation` requires also to split out `Stepping`.

All fine, I think we can move it back to `Propagation` and leave `Navigation` for the `NavigaitonDelegates` ?
The `NavigationStream` could still live in `Navigation` as it will communicate between the two, what do you think?

acts-project-service commented:
## :white_check_mark: Athena integration test results [3e0967b753560d8db1eb848900da59f166c9e45b]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928277)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928276)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928275)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928274)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928273)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928272)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928271)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928270)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928269)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928268)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928267)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928266)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928265)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928264)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928263)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928262)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928261)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928260)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928259)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928258)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928257)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928256)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928255)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928254)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928253)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928252)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928251)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928250)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928249)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928248)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928247)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928246)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42928245)||
